### PR TITLE
[RFC][MLF] Model Library Format with Multiple Modules

### PR DESCRIPTION
**Pull request by mehrdadh**
This RFC proposes adding multiple module support in Model Library Format.

cc @gromero @areusch

### CODE REVIEW

**Comment by areusch** [comment]
cc @Mousius @manupa-arm @leandron

File: rfcs/0076_mlf_with_multiple_modules.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    graph/
+      mod.graph
+  parameters/
+    mod.params
+  src/
+    mod.relay
+```
+
+One of the benefits of this approach is that create a more readable file structure which is modularized for
+each Relay build module. However, the downside is that this approach will result in more modifications in project
```
Comment (gromero) [inline rfcs/0076_mlf_with_multiple_modules.md:157]: Really can't find any issue with proposed dir. structure. Also to me it isn't such a great benefit for readability breaking down the modules in separate subdirs tbh.

**Review by gromero — COMMENTED** [review]
@mehrdadh Thanks for the RFC. Looks good!

Just some nits and I have a couple of questions for clarification and my own (better) understanding.

Cheers.

File: rfcs/0076_mlf_with_multiple_modules.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+MLF artifact is an important piece of building a microTVM project. In the first version of the MLF artifact, 
+MLF only supports a single "Relay module"/"PackedFunc" build. However, there are cases where adding
+multiple Relay modules is required. For example, corstone300 tests using multiple Relay modules was introduced by
+contributors from ARM. The goal of this RFC is to create a new version of MLF which has standard support for these cases. 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Users call `export_model_library_format` function with multiple Relay builds (output from `tvm.relay.build`) or
+single a Relay build module. We change the `export_model_library_format` API to support multiple Relay Build, however
```
[the remark this answers is not in the document]
Comment (mehrdadh) [reply rfcs/0076_mlf_with_multiple_modules.md]: done.

File: rfcs/0076_mlf_with_multiple_modules.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Users call `export_model_library_format` function with multiple Relay builds (output from `tvm.relay.build`) or
+single a Relay build module. We change the `export_model_library_format` API to support multiple Relay Build, however 
+it is backward compatible and, we handle required changes internally.
+
+Since we are including multiple modules in a single MLF file, we need to make some changes. Let's assume we
+are want to export MLF file for two Relay build modules called `mod1` and `mod2`.
```
[the remark this answers is not in the document]
Comment (mehrdadh) [reply rfcs/0076_mlf_with_multiple_modules.md]: done.

File: rfcs/0076_mlf_with_multiple_modules.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+1. We change format of `metadata.json` file. We will add `modules` as a key to JSON file which represents
+   multiple Relay build module. `modules` is a dictionary and each Relay build module is differentiated by 
+   its module name. For example:
+    ```javascript
+    {
+      'modules': {
+        'mod1': {
```
[the remark this answers is not in the document]
Comment (mehrdadh) [reply rfcs/0076_mlf_with_multiple_modules.md:36]: added more details to the RFC. Everything would be the same as before except version which will remain at top level.

File: rfcs/0076_mlf_with_multiple_modules.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        'mod2': {
+
+        },
+        ...
+      },
+      'version': XXX,
+    }
+    ```
+  2. There are multiple Relay text file generated per each Relay build. We propose to structure these files by using
+     module name and `.relay` extension to differentiate its format(E.g. {`mod1.relay`, `mod2.relay`, ...}). Similarly, for
```
[the remark this answers is not in the document]
Comment (mehrdadh) [reply rfcs/0076_mlf_with_multiple_modules.md]: By old version if you mean MLF version 6, yes we are not changing that. However, for the new version we are changing `relay.txt` to `MOD_NAME.relay`

File: rfcs/0076_mlf_with_multiple_modules.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+[drawbacks]: #drawbacks
+
+The drawback here is that we are changing generated metadata and file structure in the MLF file which means external 
+tools which are dependent to this need to update their tool. Hopefully this RFC makes it clear on what steps they need
+to take.
+
+# **Rationale and alternatives**
+An alternative way to implement this feature is to break down each Relay build module to a subdirectory and keep the
+previous format inside each sub Relay build module. Using the example of `mod1` and `mod2`, in this approach we have
+an MLF file format with structure bellow if we extract:
```
[the remark this answers is not in the document]
Comment (mehrdadh) [reply rfcs/0076_mlf_with_multiple_modules.md]: done.

File: rfcs/0076_mlf_with_multiple_modules.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+  executor-config/
+    graph/
+      mod.graph
+  parameters/
+    mod.params
+  src/
+    mod.relay
+```
+
+One of the benefits of this approach is that create a more readable file structure which is modularized for
```
[the remark this answers is not in the document]
Comment (mehrdadh) [reply rfcs/0076_mlf_with_multiple_modules.md]: should be `is that is creates` I think. I changed it to that.

File: rfcs/0076_mlf_with_multiple_modules.md (hunk 1 above)
Comment (mehrdadh) [reply rfcs/0076_mlf_with_multiple_modules.md:181]: I agree, I think it will add more complexity to the build process.

**Review by mehrdadh — COMMENTED** [review]
@gromero thanks for the review! I addressed your comments. PTAL

**Review by gromero — APPROVED** [review]
@mehrdadh Thanks for expanding the example for the new structure. LGTM.

**Pull request merged**